dsync/session.go

```diff
@@ -99,6 +102,22 @@ func (s *session) completionChanged() {
 	s.progCh <- s.prog
 }
 
+// IsFinalizedOnce will return true if the session is complete, but only the first time it is
+// called, even if multiple threads call this function at the same time
+func (s *session) IsFinalizedOnce() bool {
```

Comment:
What about making `IsFinalizedOnce` the new `Complete`?

```go
// Complete will return true if the session is complete, but only the first time it is
// called, even if multiple threads call this function at the same time
func (s *session) Complete() bool {
	if !s.prog.Complete() {
		return false
	}

	ret := false
	s.lock.Lock()
	if !s.fin {
		ret = true
		s.fin = true
	}
	defer s.lock.Unlock()
	return ret
}
```

Reply:
The confusing thing there is that is a client calls Complete multiple times, it will be false every time after the first call. Not sure if that would ever happen, but it makes the API hard to understand for future readers. I would like to have some method that is combining both the meaning of Complete, but also has a mutex-like behavior, preferably with `Once` in the name.

Reply:
Ah ok, well in that case, let's run with this.